pulsar-broker/src/main/java/org/apache/pulsar/compaction/Compactor.java

```diff
@@ -0,0 +1,205 @@
+                MessageId firstMessageId = firstMessage.getMessageId();
+                latestForKey.put(extractKey(firstMessage), firstMessageId);
+                MessageId latestMessageId = firstMessageId;
+
+                BlockingQueue<Future<RawMessage>> futures = new ArrayBlockingQueue<Future<RawMessage>>(
+                        MAX_OUTSTANDING_READS);
+                try {
+                    while (true) { // breaks with exception when there's no more messages
+                        Future<RawMessage> f = reader.readNextAsync();
+                        if (!futures.offer(f)) {
+                            RawMessage m = futures.take().get(10, TimeUnit.SECONDS);
+                            MessageId id = m.getMessageId();
+                            latestMessageId = id;
+                            String key = extractKey(m);
+                            latestForKey.put(key, id);
+                            futures.offer(f);
+                        }
+                    }
+                } catch (TimeoutException te) {
+                    // Can't read any more
+                }
+
+                futures.forEach(Compactor::cancelOrFree);
+                futures.clear();
+
+                reader.seekAsync(firstMessageId).get(10, TimeUnit.SECONDS);
+
+                LedgerHandle lh = bk.createLedger(COMPACTED_TOPIC_ENSEMBLE_SIZE,
+                                                  COMPACTED_TOPIC_ACK_QUORUM_SIZE,
+                                                  BookKeeper.DigestType.CRC32,
+                                                  COMPACTED_TOPIC_LEDGER_PASSWORD);
+                CompletableFuture<Void> lastAddFuture = CompletableFuture.completedFuture(null);
+                while (true) { // will break when it gets message matching latestMessageId
+                    Future<RawMessage> f = reader.readNextAsync();
+                    if (!futures.offer(f)) {
+                        RawMessage m = futures.take().get(10, TimeUnit.SECONDS);
+                        MessageId id = m.getMessageId();
+                        String key = extractKey(m);
+
+                        if (latestForKey.get(key).equals(id)) {
+                            final CompletableFuture<Void> bkf = new CompletableFuture<>();
+                            ByteBuf serialized = m.serialize();
+                            lh.asyncAddEntry(serialized,
+                                             (rc, ledger, eid, ctx) -> {
+                                                 if (rc != BKException.Code.OK) {
+                                                     bkf.completeExceptionally(BKException.create(rc));
+                                                 } else {
+                                                     bkf.complete(null);
+                                                 }
+                                             }, null);
+                            serialized.release();
```

Comment:
I am not sure we can `release` here. If I remember correctly, `asyncAddEntry` takes the ref and release it after callback is completed.

Reply:
asyncAddEntry calls retain at the start, so the caller should release.

https://github.com/yahoo/bookkeeper/blob/55d5c156a5100fada670700635f782dc854811ad/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java#L581